synapse/tests/test_lib_remcycle.py

```diff
@@ -0,0 +1,624 @@
+            'user_agent': 'ClownBrowser'
+        },
+        'namespace': 'vertexproject',
+    }
+    return gconfig
+
+
+def get_bad_vertex_global_config():
+    gconfig = {
+        'apis': {
+            'fake_endpoint': {
+                'doc': 'Fake endpoint',
+                'url': 'https://vertex.link/foo/bar/duck',
+                'http': {'validate_cert': False}
+            }
+        },
+        'doc': 'Grab Vertex.link stuff',
+        'http': {
+            'user_agent': 'ClownBrowser'
+        },
+        'namespace': 'vertexproject',
+    }
+    return gconfig
+
+
+def get_ipify_global_config():
+    gconfig = {
+        'apis': {
+            'jsonip': {
+                'doc': 'Get IP in a JSON blob',
+                'http': {
+                    'validate_cert': False
+                },
+                'url': 'https://api.ipify.org?format=json',
+            }
+        },
+        'doc': 'API for getting calling IP address',
+        'http': {
+            'user_agent': 'SynapseTest'
+        },
+        'namespace': 'ipify',
+    }
+    return gconfig
+
+
+def get_ipify_ingest_global_config():
+    gconfig = {
+        'apis': {
+            'jsonip': {
+                'doc': 'Get IP in a JSON blob',
+                'http': {
+                    'validate_cert': False
+                },
+                'ingests': {
+                    'ipv4': {
+                        'ingest': {
+                            "forms": [
+                                [
+                                    "inet:ipv4",
+                                    {
+                                        "var": "ip"
+                                    }
+                                ]
+                            ],
+                            "vars": [
+                                [
+                                    "ip",
+                                    {
+                                        "path": "ip"
+                                    }
+                                ]
+                            ]
+                        }
+                    }
+                },
+                'url': 'https://api.ipify.org?format=json',
+            }
+        },
+        'doc': 'API for getting calling IP address',
+        'http': {
+            'user_agent': 'SynapseTest'
+        },
+        'namespace': 'ipify',
+    }
+    return gconfig
+
+
+class NyxTest(SynTest):
+    def setUp(self):
+        self.config = {
+            "url": "http://vertex.link/api/v4/geoloc/{{someplace}}/info?domore={{domore}}&apikey={APIKEY}",
+
+            "vars": {"APIKEY": "8675309"},
+
+            "http": {
+                "method": "GET",  # this defaults to GET obvs, just making an example
+                "headers": {
+                    "x-notes-log": "stardate 1234",
+                    "token-goodness": "sekrit token"
+                },
+                "user_agent": "Totally Not a Python application."
```

Comment:
It's a Tornado thing - assuming the simplehttpclient is being used (not curl), they stamp the user agent value into the headers dictionary upon a successful connection as "User-Agent", right before writing the HTTP request.  This would overwrite a User-Agent value provided in the headers dictionary.

See https://github.com/tornadoweb/tornado/blob/8ea13db107f4d44fadd24578e693f2838012caa7/tornado/simple_httpclient.py#L314 for their under-the-hood use of the request object.